### DISCUSSION TITLE
found a buffer overflow

### DISCUSSION THREAD

**Discussion by iacore** [ggml-org/llama.cpp#819] [General]
Found this when reading the code

in `llama.cpp` when `n_dims > 2`

```
            int32_t nelements = 1;
            int32_t ne[2] = { 1, 1 };
            for (int i = 0; i < n_dims; ++i) {
                fin.read(reinterpret_cast<char *>(&ne[i]), sizeof(ne[i]));
                nelements *= ne[i];
            }
```

**Comment by sw** [comment]
#801 completely changed the loading code, `read_tensor_metadata` now checks that `1 <= n_dims <= 2`, but only after calling `file.read_raw(shard.ne.data(), sizeof(shard.ne[0]) * n_dims);`, so that's still not perfect.

**Reply by iacore** [reply]
thanks!
